akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala

```diff
@@ -334,3 +339,85 @@ private[akka] final case class Expand[In, Out, Seed](seed: In ⇒ Seed, extrapol
   final override def restart(): Expand[In, Out, Seed] =
     throw new UnsupportedOperationException("Expand doesn't support restart")
 }
+
+/**
+ * INTERNAL API
+ */
+private[akka] final case class MapAsync[In, Out](parallelism: Int, f: In ⇒ Future[Out], decider: Supervision.Decider)
```

Comment:
I think this class can be refactored to be more readable by concentrating on domain events and actions instead of the raw events and actions. For me the fundamental events are:

 - the queue becomes full or non-full (does not matter if the slots are pending or completed)
 - the front slot in the queue becomes completed
 - upstream finish or downstream cancel
 - failure

and the corresponding actions are:

 - enqueue and launch a Future
 - complete a slot
 - dequeue a completed slot

I think a refactoring around these would improve the readability.